tests/hotplug/tolerations.go

```diff
@@ -0,0 +1,124 @@
+			vm := libvmi.NewVirtualMachine(vmi, libvmi.WithRunStrategy(v1.RunStrategyAlways))
+
+			vm, err := virtClient.VirtualMachine(vm.Namespace).Create(context.Background(), vm, k8smetav1.CreateOptions{})
+			Expect(err).ToNot(HaveOccurred())
+
+			Eventually(ThisVM(vm), 360*time.Second, 1*time.Second).Should(BeReady())
+			vmi = libwait.WaitForSuccessfulVMIStart(vmi)
+
+			By("Adding tolerations")
+			tolerations := []k8sv1.Toleration{
+				{
+					Effect:            k8sv1.TaintEffectNoExecute,
+					Key:               "testTaint",
+					Operator:          k8sv1.TolerationOpExists,
+					TolerationSeconds: pointer.P(int64(300)),
+				},
+			}
+			patchSet := patch.New(patch.WithAdd("/spec/template/spec/tolerations", tolerations))
+			vm, err = patchVM(virtClient, vm, patchSet)
+			Expect(err).NotTo(HaveOccurred())
+
+			By("Ensuring the VMI has added the tolerations")
+			validateTolerations(virtClient, vmi, tolerations[0])
+
+			By("Changing the toleration")
+			tolerations[0].TolerationSeconds = pointer.P(int64(200))
+			tolerations[0].Key = "testTaint2"
+
+			patchSet = patch.New(patch.WithReplace("/spec/template/spec/tolerations", tolerations))
+			vm, err = patchVM(virtClient, vm, patchSet)
+			Expect(err).NotTo(HaveOccurred())
+
+			By("Ensuring the VMI has changed the tolerations")
+			validateTolerations(virtClient, vmi, tolerations[0])
+
+			By("Removing tolerations")
+			patchSet = patch.New(patch.WithRemove("/spec/template/spec/tolerations"))
+			vm, err = patchVM(virtClient, vm, patchSet)
+			Expect(err).NotTo(HaveOccurred())
+
+			By("Ensuring the VMI has removed the toleration")
+			Eventually(func() bool {
+				vmi, err := virtClient.VirtualMachineInstance(vmi.Namespace).Get(context.Background(), vmi.Name, k8smetav1.GetOptions{})
+				Expect(err).NotTo(HaveOccurred())
+
+				for _, vmiToleration := range vmi.Spec.Tolerations {
+					if equality.Semantic.DeepEqual(vmiToleration, tolerations[0]) {
+						return false
+					}
+				}
+
+				return true
+			}, 60*time.Second, time.Second).Should(BeTrue())
+		})
+	})
+})
+
+func patchVM(virtClient kubecli.KubevirtClient, vm *v1.VirtualMachine, patchSet *patch.PatchSet) (*v1.VirtualMachine, error) {
+	patchBytes, err := patchSet.GeneratePayload()
+	Expect(err).NotTo(HaveOccurred())
+
+	vm, err = virtClient.VirtualMachine(vm.Namespace).Patch(context.Background(), vm.Name, types.JSONPatchType, patchBytes, k8smetav1.PatchOptions{})
+	return vm, err
+}
+
+func validateTolerations(virtClient kubecli.KubevirtClient, vmi *v1.VirtualMachineInstance, toleration k8sv1.Toleration) {
+	EventuallyWithOffset(1, func() bool {
+		vmi, err := virtClient.VirtualMachineInstance(vmi.Namespace).Get(context.Background(), vmi.Name, k8smetav1.GetOptions{})
+		Expect(err).NotTo(HaveOccurred())
+
+		for _, vmiToleration := range vmi.Spec.Tolerations {
+			if equality.Semantic.DeepEqual(vmiToleration, toleration) {
+				return true
+			}
+		}
+
+		return false
+	}, 60*time.Second, time.Second).Should(BeTrue())
+}
```

Comment:
Changing this to:
```
func validateTolerations(virtClient kubecli.KubevirtClient, vmi *v1.VirtualMachineInstance, matcher gomegatypes.GomegaMatcher) {
	EventuallyWithOffset(1, func() []k8sv1.Toleration {
		vmi, err := virtClient.VirtualMachineInstance(vmi.Namespace).Get(context.Background(), vmi.Name, k8smetav1.GetOptions{})
		Expect(err).NotTo(HaveOccurred())
		return vmi.Spec.Tolerations
	}, 60*time.Second, time.Second).Should(matcher)
}
```
will allow you to use it with more flexibility.
`validateTolerations(virtClient, vmi, ContainElement(tolerations[0]))`
and the last Eventually loop (where you check the toleration deletion):
`validateTolerations(virtClient, vmi, Not(ContainElement(tolerations[0])))`

The question is: do we really need sematic.DeepEqual?

Reply:
Or just have a function that returns tolerations and you assert on caller.

Reply:
No we don't you're right, I thought there was another struct inside :+1: .